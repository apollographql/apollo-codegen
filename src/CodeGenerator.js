import {
  join,
  wrap,
} from './utilities/printing';

export default class CodeGenerator {
  constructor() {
    this.scopeStack = [];

    this.indentWidth = 2;
    this.indentLevel = 0;

    this.output = '';
  }

  pushScope(scope) {
    this.scopeStack.push(scope);
  }

  popScope() {
    return this.scopeStack.pop();
  }

  print(maybeString) {
    if (maybeString) {
      this.output += maybeString;
    }
  }

  printNewline() {
    if (this.output) {
      this.print('\n');
      this.startOfIndentLevel = false;
    }
  }

  printNewlineIfNeeded() {
    if (!this.startOfIndentLevelLevel) {
      this.printNewline();
    }
  }

  printOnNewline(maybeString) {
    if (maybeString) {
      this.printNewline();
      this.printIndent();
      this.print(maybeString);
    }
  }

  printIndent() {
    const indentation = ' '.repeat(this.indentLevel * this.indentWidth);
    this.output += indentation;
  }

  withIndent(closure) {
    if (!closure) return;

    this.indentLevel++;
    this.startOfIndentLevel = true;
    closure();
    this.indentLevel--;
  }

  withinBlock(closure) {
    this.print(' {');
    this.withIndent(closure);
    this.printOnNewline('}');
  }
}
