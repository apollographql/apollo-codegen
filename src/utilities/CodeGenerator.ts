import {CompilationContext} from '../compilation';
import {
  join,
  wrap,
} from './printing';

export default class CodeGenerator {
  private scopeStack: any[] = [];
  private indentWidth = 2;
  private indentLevel = 0;
  public output = '';
  private startOfIndentLevel = false;

  constructor(public context: CompilationContext) {

  }

  pushScope(scope: any) {
    this.scopeStack.push(scope);
  }

  popScope() {
    return this.scopeStack.pop();
  }

  print(maybeString?: string) {
    if (maybeString) {
      this.output += maybeString;
    }
  }

  printNewline() {
    if (this.output) {
      this.print('\n');
      this.startOfIndentLevel = false;
    }
  }

  printNewlineIfNeeded() {
    if (!this.startOfIndentLevel) {
      this.printNewline();
    }
  }

  printOnNewline(maybeString?: string) {
    if (maybeString) {
      this.printNewline();
      this.printIndent();
      this.print(maybeString);
    }
  }

  printIndent() {
    const indentation = ' '.repeat(this.indentLevel * this.indentWidth);
    this.output += indentation;
  }

  withIndent(closure: Function) {
    if (!closure) return;

    this.indentLevel++;
    this.startOfIndentLevel = true;
    closure();
    this.indentLevel--;
  }

  withinBlock(closure: Function, open = ' {', close = '}') {
    this.print(open);
    this.withIndent(closure);
    this.printOnNewline(close);
  }
}
