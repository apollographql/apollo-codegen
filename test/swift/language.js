import { stripIndent } from 'common-tags';

import CodeGenerator from '../../src/utilities/CodeGenerator';

import {
  classDeclaration,
  structDeclaration,
  propertyDeclaration,
  protocolDeclaration,
  protocolPropertyDeclaration
} from '../../src/swift/language';

describe('Swift code generation: Basic language constructs', function() {
  beforeEach(function() {
    this.generator = new CodeGenerator();
  });

  test(`should generate a class declaration`, function() {
    classDeclaration(this.generator, { className: 'Hero', modifiers: ['public', 'final'] }, () => {
      propertyDeclaration(this.generator, { propertyName: 'name', typeName: 'String' });
      propertyDeclaration(this.generator, { propertyName: 'age', typeName: 'Int' });
    });

    expect(this.generator.output).toBe(stripIndent`
      public final class Hero {
        public let name: String
        public let age: Int
      }
    `);
  });

  test(`should generate a struct declaration`, function() {
    structDeclaration(this.generator, { structName: 'Hero' }, () => {
      propertyDeclaration(this.generator, { propertyName: 'name', typeName: 'String' });
      propertyDeclaration(this.generator, { propertyName: 'age', typeName: 'Int' });
    });

    expect(this.generator.output).toBe(stripIndent`
      public struct Hero {
        public let name: String
        public let age: Int
      }
    `);
  });

  test(`should generate nested struct declarations`, function() {
    structDeclaration(this.generator, { structName: 'Hero' }, () => {
      propertyDeclaration(this.generator, { propertyName: 'name', typeName: 'String' });
      propertyDeclaration(this.generator, { propertyName: 'friends', typeName: '[Friend]' });

      structDeclaration(this.generator, { structName: 'Friend' }, () => {
        propertyDeclaration(this.generator, { propertyName: 'name', typeName: 'String' });
      });
    });

    expect(this.generator.output).toBe(stripIndent`
      public struct Hero {
        public let name: String
        public let friends: [Friend]

        public struct Friend {
          public let name: String
        }
      }
    `);
  });

  test(`should generate a protocol declaration`, function() {
    protocolDeclaration(this.generator, { protocolName: 'HeroDetails', adoptedProtocols: ['HasName'] }, () => {
      protocolPropertyDeclaration(this.generator, { propertyName: 'name', typeName: 'String' });
      protocolPropertyDeclaration(this.generator, { propertyName: 'age', typeName: 'Int' });
    });

    expect(this.generator.output).toBe(stripIndent`
      public protocol HeroDetails: HasName {
        var name: String { get }
        var age: Int { get }
      }
    `);
  });
});
