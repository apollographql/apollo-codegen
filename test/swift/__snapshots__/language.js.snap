// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swift code generation: Basic language constructs should handle multi-line descriptions 1`] = `
"// A hero
public struct Hero {
  public let name: String
  // A multiline comment
  // on the hero's name.
}"
`;
